### PR TITLE
Dev: Implemented http proxy authorization

### PR DESCRIPTION
**Pull request by bvn13**
Usage example:
```java
        try {
            ApiContextInitializer.init();

            // Create the TelegramBotsApi object to register your bots
            TelegramBotsApi botsApi = new TelegramBotsApi();

            DefaultBotOptions botOptions = ApiContext.getInstance(DefaultBotOptions.class);
            if (NEED_AUTHORIZATION) {
                CredentialsProvider credsProvider = new BasicCredentialsProvider();
                credsProvider.setCredentials(
                        new AuthScope(PROXY_HOST, PROXY_PORT),
                        new UsernamePasswordCredentials(PROXY_USER, PROXY_USER_PASS));

                HttpHost httpHost = new HttpHost(PROXY_HOST, PROXY_PORT);

                RequestConfig requestConfig = RequestConfig.custom().setProxy(httpHost).setAuthenticationEnabled(true).build();
                botOptions.setRequestConfig(requestConfig);
                botOptions.setCredentialsProvider(credsProvider);
                botOptions.setHttpProxy(httpHost);
            }

            // Register your newly created AbilityBot
            TelegramBot bot = new TelegramBot(TOKEN, BOT_NAME, botOptions);
            botsApi.registerBot(bot);

            //session.setOptions(botOptions);
        } catch (TelegramApiException e) {
            e.printStackTrace();
        }
```

### CODE REVIEW

File: telegrambots/src/main/java/org/telegram/telegrambots/bots/DefaultAbsSender.java [hunk 1]
```diff
@@ -87,6 +100,29 @@ public final DefaultBotOptions getOptions() {
         return options;
     }
 
+    protected CloseableHttpClient createHttpClient() {
+        CloseableHttpClient localClient = null;
+
+        if (options.getCredentialsProvider() != null) {
```
Comment (IgnatBeresnev) [inline telegrambots/src/main/java/org/telegram/telegrambots/bots/DefaultAbsSender.java]: Why such order?

```
if(not null) {
} else (not not null) {
}
```
in my opinion is more difficult to understand than
```
if(null){
} else (not null) {
}
```

Comment (bvn13) [reply telegrambots/src/main/java/org/telegram/telegrambots/bots/DefaultAbsSender.java]: As for me it does not matter.

Comment (bvn13) [reply telegrambots/src/main/java/org/telegram/telegrambots/bots/DefaultAbsSender.java]: What if it was

```
if (not null) {}
else if (null) {}
```

**Comment by bvn13** [comment]
@vellrya попробуйте использовать этот прокси в браузере и в десктопном клиенте телеграма. у меня есть ощущение, что прокси не работает

**Comment by vellrya** [comment]
@bvn13 можно ли как-то связаться с вами в телеграме, пожалуйста? Мой ник: @ryazantsev_val
Хотелось бы увидеть пример использования, потому что так, как вы указали в описании, запустить бота не получилось

**Comment by rubenlagus** [comment]
Please, keep discussion in English.

**Comment by Bogus-Coder** [comment]
What about socks5-proxy support?

**Comment by bvn13** [comment]
@Bogus-Coder I didn't find a glue for easiest using socks5-proxy yet.

**Comment by dygy** [comment]
What about some Tor?)

**Comment by carlopantaleo** [comment]
@bvn13 What if a proxy doesn't need authorisation? May one just skip setting the ``CredentialProvider``?
```java
// botOptions.setCredentialsProvider(credsProvider);
botOptions.setHttpProxy(httpHost);
```

**Comment by bvn13** [comment]
@carlopantaleo yes, just skip it. Here is an example

```java
// Create the TelegramBotsApi object to register your bots
TelegramBotsApi botsApi = new TelegramBotsApi();

// Set up Http proxy
DefaultBotOptions botOptions = ApiContext.getInstance(DefaultBotOptions.class);

HttpHost httpHost = new HttpHost(PROXY_HOST, PROXY_PORT);

RequestConfig requestConfig = RequestConfig.custom().setProxy(httpHost).setAuthenticationEnabled(false).build();
botOptions.setRequestConfig(requestConfig);
botOptions.setHttpProxy(httpHost);

// Register your newly created AbilityBot
MyBot bot = new MyBot(BOT_TOKEN, BOT_NAME, botOptions);

botsApi.registerBot(bot);
```

**Comment by tghns** [comment]
Maybe it's a stupid question, but I'm new to GitHub, so could you please explain, how I can add your implementation to my project?  

I guess, I should somehow add the dev branch to the project, but really don't know how.

**Comment by bvn13** [comment]
@tghns if you are russian, you could read my article here https://bvn13.tk/posts/java-telegram-bot-using-http-proxy

otherwise, you may clone my repo https://github.com/bvn13/TelegramBots, build and install it with maven `mvn install`. And use it
_for maven_:
```xml
 <!--TelegramAPI-->
    <dependency>
        <groupId>org.telegram</groupId>
        <artifactId>telegrambots-abilities</artifactId>
        <version>3.6</version>
    </dependency>
``
_or gradle_:
```
    // telegram bot api
    compile 'org.telegram:telegrambots:3.6'
```

Note to change priority of repos in gradle - local must be first:
```
repositories {
    mavenLocal()
    mavenCentral()
    jcenter()
}

buildscript {
    repositories {
        mavenLocal()
        mavenCentral()
        jcenter()
   }
}
```

**Comment by dygy** [comment]
botOptions.setHttpProxy(httpHost);
can't resolve setHttpProxy . Can you help me?

**Comment by dygy** [comment]
and I really like to see some examples of the full main method. It might be some of my mistakes. Thx for helping,

**Comment by bvn13** [comment]
@dygy You didn't use my repo. Check my explanation above. 
1) use my repo
2) build and install TelegramBots from my repo
3) make sure you use it in your project

**Comment by bvn13** [comment]
@dygy it's the same for gradle. Check if gradle use it. Place mavenLocal() on top of repolist:

```
repositories {
    mavenLocal()
    mavenCentral()
    jcenter()
}

buildscript {
    repositories {
        mavenLocal()
        mavenCentral()
        jcenter()
   }
}
```

**Comment by dygy** [comment]
still 
`Error:(28, 23) java: cannot find symbol
  symbol:   method setHttpProxy(org.apache.http.HttpHost)
  location: variable botOptions of type org.telegram.telegrambots.bots.DefaultBotOptions`

**Comment by bvn13** [comment]
@rubenlagus should I write a wiki page for using proxy?

**Comment by rubenlagus** [comment]
@bvn13 If you write it, it would be great. You can add the changes to this pull request and I'll merge them together with the release.

My idea is to merge everything today and get the version ready (today or tomorrow)

**Comment by tghns** [comment]
Thank you, it works

**Comment by bvn13** [comment]
@rubenlagus I wrote and added wiki page, but I actually don't know if it would be linked correctly into sidebar. Could you check it?

**Comment by jakeostis** [comment]
for those who encountered difficulties due to blocking Telegram in Russia - try to register your implemented bot with this proxy settings (for socks5 protocol):
System.getProperties().put( "socksProxyHost", "host" );
System.getProperties().put( "socksProxyPort", "port" );

if proxy server requires authentication add:

        Authenticator.setDefault(new Authenticator(){
            protected PasswordAuthentication getPasswordAuthentication(){
                return new PasswordAuthentication("username", "password".toCharArray());
            }
        });

**Comment by bvn13** [comment]
@jakeostis what if my bot works with other services placed in local network? I want my bot to use a proxy only for telegram servers, not for my local network or other services.

**Comment by jakeostis** [comment]
@bvn13 i absolutely agree. i have edited my comment. Hence, this solution is not flexible for everyone and implementation proxy in code will be much better

**Comment by IgnatBeresnev** [comment]
@jakeostis @bvn13 I'm sure you don't have to set the system property to get it working. Try running your app with `java ... -DsocksProxyHost=host -DsocksProxyPort=port`

**Pull request merged**